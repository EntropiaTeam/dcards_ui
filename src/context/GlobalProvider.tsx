import React from 'react';
import {
  CardProvider,
  CategoriesProvider,
  TextProvider,
  OrderProvider,
  CustomImageProvider
} from '.';
import { NavigationProvider } from './navigation';

export const GlobalProvider: React.FunctionComponent = ({ children }) => (
  <NavigationProvider>
    <CardProvider>
      <CategoriesProvider>
        <TextProvider>
          <CustomImageProvider>
            <OrderProvider>
              {children}
            </OrderProvider>
          </CustomImageProvider>
        </TextProvider>
      </CategoriesProvider>
    </CardProvider>
  </NavigationProvider>
);
