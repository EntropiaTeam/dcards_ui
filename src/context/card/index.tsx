import React, {
  createContext, useContext, useReducer, FC
} from 'react';
import { CardActionsType, UseCardContext } from './types/CardTypes';
import CardReducer, { initialState } from './reducer/CardReducer';
import { CardActions } from './actions/CardActions';

// Create Context
export const CardContext = createContext(initialState);
export const CardActionsContext = createContext({} as CardActionsType);
export const useCardContext = (): UseCardContext => (
  { state: useContext(CardContext), actions: useContext(CardActionsContext) }
);

export const CardProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CardReducer, initialState);

  // Actions
  const actions: CardActionsType = {
    setCardToEdit: CardActions.setCardToEdit(dispatch)
  };

  return (
    <CardContext.Provider value={state}>
      <CardActionsContext.Provider value={actions}>
        {children}
      </CardActionsContext.Provider>
    </CardContext.Provider>
  );
};
