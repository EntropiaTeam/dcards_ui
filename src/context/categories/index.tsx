import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CategoriesActionsType, UseCategoriesContext } from './types/CategoriesTypes';
import CategoriesReducer, { initialState } from './reducer/CategoriesReducer';
import { CategoriesActions } from './actions/CategoriesActions';

// Create Context
export const CategoriesContext = createContext(initialState);
export const CategoriesActionsContext = createContext({} as CategoriesActionsType);
export const useCategoriesContext = (): UseCategoriesContext => ({
  state: useContext(CategoriesContext), actions: useContext(CategoriesActionsContext)
});

export const CategoriesProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CategoriesReducer, initialState);

  // Actions
  const actions: CategoriesActionsType = useMemo(() => ({
    setCategories: CategoriesActions.setCategories(dispatch)
  }), []);

  return (
    <CategoriesContext.Provider value={state}>
      <CategoriesActionsContext.Provider value={actions}>
        {children}
      </CategoriesActionsContext.Provider>
    </CategoriesContext.Provider>
  );
};
