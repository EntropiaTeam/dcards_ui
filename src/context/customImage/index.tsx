import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CustomImageActionsType, UseCustomImageContext } from './types/CustomImageTypes';
import { CustomImageActions } from './actions/CustomImageActions';
import CustomImageReducer, { initialState } from './reducer/CustomImageReducer';

// Create Context
export const CustomImageContext = createContext(initialState);
export const CustomImageActionsContext = createContext({} as CustomImageActionsType);
export const useCustomImageContext = (): UseCustomImageContext => (
  { state: useContext(CustomImageContext), actions: useContext(CustomImageActionsContext) }
);

export const CustomImageProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CustomImageReducer, initialState);
  // Actions
  const actions: CustomImageActionsType = useMemo(() => ({
    setErrors: CustomImageActions.setErrors(dispatch),
    setUserImage: CustomImageActions.setUserImage(dispatch),
    setCropper: CustomImageActions.setCropper(dispatch),
    setAgreement: CustomImageActions.setAgreement(dispatch),
    setAgreementTranslation: CustomImageActions.setAgreementTranslation(dispatch),
    setCroppedImage: CustomImageActions.setCroppedImageValue(dispatch),
    setCropperBackImageURL: CustomImageActions.setCropperBackImageValue(dispatch),
    setRotationDegree: CustomImageActions.setRotationDegree(dispatch),
    setZoomRatio: CustomImageActions.setZoomRatio(dispatch),
    setCroppedCanvasData: CustomImageActions.setCroppedCanvasData(dispatch),
    resetCustomImageState: CustomImageActions.resetCustomImageState(dispatch)
  }), [dispatch]);

  return (
    <CustomImageContext.Provider value={state}>
      <CustomImageActionsContext.Provider value={actions}>
        {children}
      </CustomImageActionsContext.Provider>
    </CustomImageContext.Provider>
  );
};
