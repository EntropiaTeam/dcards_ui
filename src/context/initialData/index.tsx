import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { UseInitialDataContext, InitialDataActionsType } from './types/InitialDataTypes';
import InitialDataReducer, { initialState as initDatainitialState } from './reducer/InitialDataReducer';
import { InitialDataActions } from './actions/InitialDataActions';

import {
  useCategoriesContext,
  useCustomImageContext
} from '..';

// Create Context
export const InitialDataContext = createContext(initDatainitialState);
export const InitialDataActionsContext = createContext({} as InitialDataActionsType);
export const useInitialDataContext = (): UseInitialDataContext => ({
  state: useContext(InitialDataContext), actions: useContext(InitialDataActionsContext)
});

export const InitialDataProvider: FC = ({ children }) => {
  const { actions: { setCategories } } = useCategoriesContext();
  const { actions: { setAgreementTranslation } } = useCustomImageContext();
  const [initDatastate, dispatchInitData] = useReducer(InitialDataReducer, initDatainitialState);

  // Actions
  const actions: InitialDataActionsType = useMemo(() => ({
    loadInitialData: InitialDataActions.loadInitialData(
      dispatchInitData,
      setCategories,
      setAgreementTranslation
    )
  }), [setCategories, setAgreementTranslation]);

  return (
    <InitialDataContext.Provider value={initDatastate}>
      <InitialDataActionsContext.Provider value={actions}>
        {children}
      </InitialDataActionsContext.Provider>
    </InitialDataContext.Provider>
  );
};
