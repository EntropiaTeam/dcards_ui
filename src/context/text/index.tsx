import React, {
  createContext, useContext, useMemo, useReducer, FC
} from 'react';
import TextReducer, { initialState } from './reducer/TextReducer';
import { TextActionsType, UseTextContext } from './types/TextTypes';
import { TextActions } from './actions/TextActions';

// Create Context
export const TextContext = createContext(initialState);
export const TextActionsContext = createContext({} as TextActionsType);
export const useTextContext = (): UseTextContext => ({
  state: useContext(TextContext),
  actions: useContext(TextActionsContext)
});

export const TextProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(TextReducer, initialState);

  // Actions
  const actions: TextActionsType = useMemo(() => ({
    setCustomText: TextActions.setCustomText(dispatch),
    validateTextHeight: TextActions.validateTextHeight(dispatch),
    setMaxTextareaHeight: TextActions.setMaxTextareaHeight(dispatch),
    setCurrentTextareaHeight: TextActions.setCurrentTextareaHeight(dispatch),
    resetCustomTextState: TextActions.resetCustomTextState(dispatch)
  }), []);

  return (
    <TextContext.Provider value={state}>
      <TextActionsContext.Provider value={actions}>
        {children}
      </TextActionsContext.Provider>
    </TextContext.Provider>
  );
};
