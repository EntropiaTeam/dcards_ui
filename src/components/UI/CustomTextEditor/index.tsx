import React, {
  ChangeEvent, forwardRef, useCallback, useState, useEffect
} from 'react';
import { useTranslation } from 'react-i18next';
import TextareaAutosize from 'react-textarea-autosize';
import { makeStyles, Theme } from '@material-ui/core';
import { useTextContext } from '../../../context';
import useAppConfig from '../../../hooks/useAppConfig';
import { AppConfig } from '../../../utils/AppConfig';

const useStyles = makeStyles((theme: Theme) => ({
  textArea: {
    width: '100%',
    minHeight: '48px',
    marginTop: ({ topOffset }: StyleProps) => (topOffset),
    padding: 0,
    border: 'none',
    outline: 'none',
    resize: 'none',
    textAlign: 'center',
    fontFamily: 'inherit',
    backgroundColor: 'transparent',
    fontSize: 'inherit',
    lineHeight: ({ lineHeightSize }: StyleProps) => `${lineHeightSize}px`,
    fontWeight: 'inherit',
    color: 'inherit',
    cursor: 'pointer',
    overflow: 'hidden',
    lineBreak: 'normal',
    '&::placeholder': {
      color: '#C8102E',
      fontFamily: theme.typography.button.fontFamily,
      fontSize: '16px',
      textDecoration: 'underline'
    },
    transform: ({ desktopFitScale }: StyleProps) => `scale(${desktopFitScale || 1})`,
    transformOrigin: '50% 0'
  },
  textAreaWrapper: {
    margin: '0 auto',
    width: ({ width }) => width,
    height: ({
      desktopFitScale, currentTextareaHeight
    }: StyleProps) => `${desktopFitScale ? (desktopFitScale * currentTextareaHeight) : currentTextareaHeight}px`
  }
}));

type Props = {
  isMobileView?: boolean;
  placeholderText?: string;
  desktopFitScale?: number;
  textareaHeight: number;
  pageHeight: number;
  width?: string;
  topOffset?: string;
  tabIndex?: number;
  blurHandler?: () => void;
};

type StyleProps = {
  desktopFitScale?: number;
  topOffset: string;
  currentTextareaHeight: number;
  lineHeightSize: number;
  width: string;
};

const CustomTextEditor = forwardRef<HTMLTextAreaElement, Props>(({
  placeholderText,
  isMobileView,
  textareaHeight,
  pageHeight,
  desktopFitScale,
  width,
  topOffset,
  tabIndex = 0,
  blurHandler
}, editorRef) => {
  const [pasting, setPasting] = useState(false);
  const [isTextEditorFocused, setIsTextEditorFocused] = useState(false);
  const [combinedEditorRef] = useState(
    editorRef as unknown as React.MutableRefObject<HTMLInputElement>
  );
  const {
    state: {
      isTextHeightValid, maxTextareaHeight, currentTextareaHeight
    },
    actions: {
      setCustomText, setMaxTextareaHeight, setCurrentTextareaHeight
    }
  } = useTextContext();
  const { appConfig, cardConfig } = useAppConfig();
  const { font, lineHeight: { sizeRatio } } = cardConfig;
  const lineHeightSize = Number(font.size) * sizeRatio;
  const maxTextareaLength = (37 * (Math.floor(maxTextareaHeight / lineHeightSize)));
  const { t } = useTranslation();
  const localizedPlaceholder = placeholderText && t(placeholderText);
  const computedPixelHeight = AppConfig.percentsToPixelsHeight(pageHeight, textareaHeight);
  const classes = useStyles({
    desktopFitScale,
    currentTextareaHeight,
    lineHeightSize,
    width: width || '100%',
    topOffset: topOffset || '0px'
  });

  const textareaHeightHandler = useCallback((height: number): void => {
    setCurrentTextareaHeight(height);
    setMaxTextareaHeight(computedPixelHeight);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [computedPixelHeight]);

  const focusOutTextarea = useCallback((): void => {
    setIsTextEditorFocused(false);
    if (blurHandler) {
      blurHandler();
    }
  }, [blurHandler]);

  const focusOnTextarea = useCallback((): void => {
    setIsTextEditorFocused(true);
  }, []);

  useEffect(() => {
    if (combinedEditorRef
        && combinedEditorRef.current
        && isMobileView
    ) {
      const bottomOffset = isTextEditorFocused ? currentTextareaHeight * 1.5 : 0;
      combinedEditorRef.current.style.marginBottom = `${bottomOffset}px`;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentTextareaHeight, isTextEditorFocused]);

  const customTextHandler = (event: ChangeEvent<HTMLTextAreaElement>): void => {
    const { value } = event.target;

    if (pasting) {
      const pastedValue = value
        .replace(/(^\s*)|(\s*$)/gi, '') // removes leading and trailing spaces
        .replace(/[ ]{2,}/gi, ' ') // replaces multiple spaces with one space
        .replace(/\n +/, '\n') // removes spaces after newlines
        .replace(/(\r\n|\n|\r)/gm, ''); // removes new lines and tabs

      let pastedSubstringValue = pastedValue.substring(0, maxTextareaLength);

      if (pastedValue.length >= maxTextareaLength
          && combinedEditorRef
          && combinedEditorRef.current
      ) {
        pastedSubstringValue += '\n';
        combinedEditorRef.current.value = pastedSubstringValue;
      }

      setCustomText(pastedSubstringValue);
      setPasting(false);
    }

    if (!pasting) {
      setCustomText(value);
    }
  };

  const handlePaste = (): void => {
    setPasting(true);
    if (combinedEditorRef && combinedEditorRef.current) {
      textareaHeightHandler(Number(combinedEditorRef.current.clientHeight));
    }
  };

  const placeholder = !isTextEditorFocused ? localizedPlaceholder : '';

  return (
    <div className={classes.textAreaWrapper}>
      <TextareaAutosize
        ref={editorRef}
        maxLength={isTextHeightValid ? appConfig.maxTextareaLength : 0}
        className={classes.textArea}
        placeholder={placeholder}
        onChange={customTextHandler}
        onHeightChange={(height) => textareaHeightHandler(height)}
        onBlur={focusOutTextarea}
        onFocus={focusOnTextarea}
        onPaste={handlePaste}
        tabIndex={tabIndex}
      />
    </div>
  );
});

export default CustomTextEditor;
