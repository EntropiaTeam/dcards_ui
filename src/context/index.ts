import { CardProvider, useCardContext } from './card';
import { CategoriesProvider, useCategoriesContext } from './categories';
import { TextProvider, useTextContext } from './text';
import { OrderProvider, useOrderContext } from './order';
import { CustomImageProvider, useCustomImageContext } from './customImage';
import { NavigationProvider, useNavigationContext } from './navigation';

export {
  CardProvider,
  useCardContext,
  CategoriesProvider,
  useCategoriesContext,
  TextProvider,
  useTextContext,
  OrderProvider,
  useOrderContext,
  CustomImageProvider,
  useCustomImageContext,
  NavigationProvider,
  useNavigationContext
};
