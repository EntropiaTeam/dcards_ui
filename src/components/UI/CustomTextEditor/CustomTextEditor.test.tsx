import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import CustomTextEditor from '.';
import { TextProvider } from '../../../context/text';

interface IntlMock {
  useTranslation: () => void;
}

interface TranslationMock {
  t: () => void;
}

interface TextAreaProps {
  onHeightChange: () => void;
}

interface ElementProps {
  props: () => void;
}

jest.mock('react-i18next', (): IntlMock => ({
  useTranslation: (): TranslationMock => ({
    t: (key: string): string => key
  })
}));

describe('<CustomTextEditor />', () => {
  const blurHandlerMock = jest.fn();
  const props = {
    blurHandler: blurHandlerMock,
    placeholderText: 'Test placeholder',
    tabIndex: 2,
    maxLength: 370,
    pageHeight: 200,
    textareaHeight: 400
  };

  const wrapper = mount(
    <TextProvider>
      <CustomTextEditor {...props} />
    </TextProvider>
  );
  const textAreaComponent = wrapper.childAt(0).find('textarea').first();

  it('should call bloor function', () => {
    act(() => {
      textAreaComponent.props().onBlur();
    });
    expect(blurHandlerMock).toHaveBeenCalledTimes(1);
  });

  it('should set correct props in deep', () => {
    const settedPlaceholder = textAreaComponent.getDOMNode().getAttribute('placeholder');
    const tabIndex = textAreaComponent.getDOMNode().getAttribute('tabIndex');
    expect(settedPlaceholder).toBe(props.placeholderText);
    expect(tabIndex).toBe(String(props.tabIndex));
  });

  it('should track change height', () => {
    const parentCompnent: ElementProps = textAreaComponent.parent();
    const parentProps: TextAreaProps = parentCompnent.props();
    const node = textAreaComponent.getDOMNode();
    const original = node.getAttribute('maxLength');
    act(() => {
      parentProps.onHeightChange(500);
    });
    const compared = node.getAttribute('maxLength');
    expect(original).toBeTruthy();
    expect(compared).toBeTruthy();
    expect(Number(compared)).toBeGreaterThanOrEqual(Number(original));
  });

  it('should render CustomTextEditor component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
