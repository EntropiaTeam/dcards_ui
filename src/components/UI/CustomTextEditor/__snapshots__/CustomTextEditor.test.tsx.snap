// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomTextEditor /> should render CustomTextEditor component correctly 1`] = `ReactWrapper {}`;
